alibi_detect/cd/pytorch/lsdd_online.py

```diff
@@ -0,0 +1,201 @@
+            The number of bootstrap simulations used to configure the thresholds. The larger this is the
+            more accurately the desired ERT will be targeted. Should ideally be at least an order of magnitude
+            larger than the ert.
+        n_kernel_centers
+            The number of reference samples to use as centers in the Gaussian kernel model used to estimate LSDD.
+            Defaults to 2*window_size.
+        lambda_rd_max
+            The maximum relative difference between two estimates of LSDD that the regularization parameter
+            lambda is allowed to cause. Defaults to 0.2 as in the paper.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'. Only relevant for 'pytorch' backend.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            ert=ert,
+            window_size=window_size,
+            preprocess_x_ref=preprocess_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            n_bootstraps=n_bootstraps,
+            n_kernel_centers=n_kernel_centers,
+            lambda_rd_max=lambda_rd_max,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'pytorch'})
+
+        # set backend
+        if device is None or device.lower() in ['gpu', 'cuda']:
+            self.device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        self._configure_normalization()
+
+        # initialize kernel
+        if sigma is None:
+            self.x_ref = torch.from_numpy(self.x_ref).to(self.device)
+            self.kernel = GaussianRBF()
+            _ = self.kernel(self.x_ref, self.x_ref, infer_sigma=True)
+        else:
+            sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma, np.ndarray) else None
+            self.kernel = GaussianRBF(sigma)
+
+        self._configure_kernel_centers()
+        self._configure_thresholds()
+        self._initialise()
+
+    def _configure_normalization(self):
+        x_ref_means = self.x_ref.mean(axis=0)
+        x_ref_stds = self.x_ref.std(axis=0)
+        self._normalize = lambda x: (x - x_ref_means)/(x_ref_stds)
+        self.x_ref = self._normalize(self.x_ref)
+
+    def _configure_kernel_centers(self):
+        "Set aside reference samples to act as kernel centers"
+        perm = torch.randperm(self.n)
+        self.c_inds, self.non_c_inds = perm[:self.n_kernel_centers], perm[self.n_kernel_centers:]
+        self.kernel_centers = self.x_ref[self.c_inds]
```

Comment:
No check for `np.unique(self.kernel_centers.numpy(), axis=0).shape[0] < self.n_kernel_centers`?